### ISSUE TITLE
Official container doesn't run in non-Docker runtime on default port

### ISSUE THREAD

**Issue by thw0rted** [pgadmin-org/pgadmin4#6101]
**Describe the bug**

The official container (`docker.io/dpage/pgadmin4`) listens on port 80 by default, with the primary server application owned by user `pgadmin` (not `root`).

In Docker, this is supported [due to a change made a couple of years ago](https://github.com/moby/moby/pull/41030) to allow non-privileged processes to use what would otherwise be considered privileged ports.  There are other container environments, though, and some of them still enforce typical Linux port restrictions.  In these other environments, running the container without setting `PGADMIN_LISTEN_PORT` to a "high" (nonprivileged) value will fail.

**To Reproduce**

1. Run the image in a non-Docker environment, like CloudFoundry
2. Startup fails with an error message that it cannot bind to port 80

**Expected behavior**

The container should not try to bind a normally-privileged port by default.  Since the runtime environment has to explicitly map any exposed ports anyway, there's no reason to use port 80 of the container instead of e.g. 8080.

**Error message**

```
NOTE: Configuring authentication for SERVER mode.
pgAdmin 4 - Application Initialisation
======================================
[INFO] Starting gunicorn 20.1.0
[ERROR] Retrying in 1 second.
[ERROR] Retrying in 1 second.
[ERROR] Retrying in 1 second.
[ERROR] Retrying in 1 second.
[ERROR] Retrying in 1 second.
[ERROR] Can't connect to ('0.0.0.0', 80)
```

**Desktop (please complete the following information):**
 - OS: CloudFoundry
 - Version: 2.12.19
 - Mode: Server
 - Package type: container

**Comment by thw0rted** [comment]
I found out about the Docker situation from [this blog post](https://medium.com/@olivier.gaumond/why-am-i-able-to-bind-a-privileged-port-in-my-container-without-the-net-bind-service-capability-60972a4d5496), which might be a useful reference for others encountering this issue.

**Comment by dpage** [comment]
Hi,

This is not something we can change now without potentially breaking the majority of existing deployments. You can use an alternative port that will work in your deployment by setting the PGADMIN_LISTEN_PORT environment variable; see https://www.pgadmin.org/docs/pgadmin4/6.21/container_deployment.html#environment-variables for more info.

**Comment by thw0rted** [comment]
The problem with `PGADMIN_LISTEN_PORT` is that the container's Dockerfile still `EXPOSE`s ports 80 and 443 (only).  From [the Dockerfile spec](https://docs.docker.com/engine/reference/builder/#expose):

> The EXPOSE instruction does not actually publish the port. It functions as a type of documentation between the person who builds the image and the person who runs the container, about which ports are intended to be published. To actually publish the port when running the container, use the -p flag on docker run to publish and map one or more ports, or the -P flag to publish all exposed ports and map them to high-order ports.

So basically, when you allow `PGADMIN_LISTEN_PORT` to change the port behavior at runtime, the `EXPOSE` "documentation" becomes inaccurate.  This is a problem when downstream tooling (like, again, CloudFoundry) expects that metadata to be accurate.  This leaves users in the unfortunate situation where the container will not run at all in its default configuration, and configuring the container so it can run (by changing the port) results in downstream tooling connecting to the wrong port.  CloudFoundry requires some pretty arcane hoop-jumping to reconfigure an application when its actual runtime listen-port is not the one described in the Dockerfile.

**Comment by thw0rted** [comment]
This week I've reading a bit more on old https://github.com/moby/moby/ issues, and I kept seeing people arguing that containers are supposed to be "reproducible" -- basically arguing *against* configurable runtime behavior.

I thought I was going to fix this by building a Dockerfile `FROM` your image, then `EXPOSE`ing the port I actually run on.  The problem with that is you can never undo `EXPOSE`, or otherwise "reset" most parent container metadata, so I was actually making a container that `EXPOSE`s 80, 443, and 8080 -- and CloudFoundry still routes everything to the lowest `EXPOSE`d port.

People posted hacky workarounds, like tools that serialize a container with `docker save`, pipe the entire metadata JSON through a script that removes some values, then rehydrate the result with `docker load`, or they suggested copying the original Dockerfile minus the instructions you don't want (though of course there's no good way to maintain that going forward...).  Eventually I gave up and used the Cloud Foundry REST API to rewire the routes, so my personal problem is solved, but I think there's still an impedance mismatch between the Docker community's expectations and how this image is designed.

(As an aside: Docker allows you to map between internal and external ports at will using `-p xxx:yyy`, so I'm curious what the original motivation was for allowing reconfiguration of the internal port by environment variable.  I guess that would allow you to install additional services inside the container, but isn't the point of `compose` / k8s to let containers do one thing well, then glue them together?)

**Comment by dpage** [comment]
Configurable runtime behaviour is unavoidable; differing security policies in different orgs are an obvious example where people need to be able to configure in different ways (e.g. in some environments people will use Postfix in the container for ease of configuration, whilst others will disable it and use an external MTA). Some users also prefer to handle things like HTTPS in the pgAdmin container, whilst others will use a reverse proxy of some sort to do it.

Our general preference is to give people the option to run how best suits them, rather than get drawn into the whole argument around configurability (or lack of). Some folks simply want to run pgAdmin in a container on their laptop after all.

Can you confirm that CloudFoundry always exposes the lowest port, and not the **first** port? How does it handle containers that expose multiple ports as part of the same application?

**Comment by thw0rted** [comment]
[The docs say](https://docs.cloudfoundry.org/devguide/deploy-apps/push-docker.html) that "Cloud Foundry supports only one exposed port on the image".

I'm not sure where I saw the bit about picking the lowest (vice first-declared) port, but it might have been [this comment on SO](https://stackoverflow.com/questions/67951182/port-setting-of-docker-images-in-cloudfoundry#comment120136177_67951182).  I did a bit more googling and found [this post on their forums](https://community.pivotal.io/s/question/0D50e00006sVi7BCAS/tas-29-docker-image-deployment-error-connecting-to-port?language=en_US) where an employee says it's the lowest, so that's probably good enough for me.  (I'd dig around in the code to be sure, if I had time, but honestly we're in the process of moving out of Cloud Foundry anyway so I don't want to burn more cycles on this.)
